src/Microsoft.DotNet.Interactive.CSharp/CSharpKernel.cs

```diff
@@ -286,19 +290,23 @@ public async Task HandleAsync(SubmitCode submitCode, KernelInvocationContext con
             }
         }
 
+        public Task HandleAsync(ChangeWorkspaceDirectory command, KernelInvocationContext context)
+        {
+            _workspaceDirectory = command.WorkspaceDirectory;
+            return Task.CompletedTask;
+        }
+
         private async Task RunAsync(
             string code,
             CancellationToken cancellationToken = default,
             Func<Exception, bool> catchException = default)
         {
-            var currentDirectory = Directory.GetCurrentDirectory();
-            if (_currentDirectory != currentDirectory)
-            {
-                _currentDirectory = currentDirectory;
-                ScriptOptions = ScriptOptions.WithMetadataResolver(
-                    CachingMetadataResolver.Default.WithBaseDirectory(
-                        _currentDirectory));
-            }
+            if (_workspaceDirectory == null)
+                _workspaceDirectory = Directory.GetCurrentDirectory();
+
+            ScriptOptions = ScriptOptions
+                .WithMetadataResolver(CachingMetadataResolver.Default.WithBaseDirectory(_workspaceDirectory))
+                .WithSourceResolver(new SourceFileResolver(ImmutableArray<string>.Empty, _workspaceDirectory));
```

Comment:
I'd like to avoid setting the `_workspaceDirectory` variable unless we explicitly enter the `HandleAsync()` method that you added.  I'd prefer something like this:

``` csharp
var workingDirectory = _workspaceDirectory ?? Directory.GetCurrentDirectory();

ScriptOptions = ScriptOptions
    .WithMetadataResolver(CachingMetadataResolver.Default.WithBaseDirectory(workingDirectory))
    .WithSourceResolver(new SourceFileResolver(ImmutableArray<string>.Emtpy, workingDirectory));
```

Same change applied to the F# kernel, too.

Reply:
Yes, I agree,
"_workspaceDirectory" should not mutate magically as a side-effect.